File: app/src/main/java/com/boardgamegeek/ui/adapter/ThreadRecyclerViewAdapter.java
package com.boardgamegeek.ui.adapter;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.boardgamegeek.R;
import com.boardgamegeek.databinding.RowThreadArticleBinding;
import com.boardgamegeek.entities.ForumEntity.ForumType;
import com.boardgamegeek.extensions.TextViewUtils;
import com.boardgamegeek.ui.ArticleActivity;
import com.boardgamegeek.ui.loader.ThreadSafeResponse;
import com.boardgamegeek.ui.model.Article;
import com.boardgamegeek.ui.widget.TimestampView;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

public class ThreadRecyclerViewAdapter extends RecyclerView.Adapter<ThreadRecyclerViewAdapter.ArticleViewHolder> {
	private final int threadId;
	private final String threadSubject;
	private final int forumId;
	private final String forumTitle;
	private final int objectId;
	private final String objectName;
	private final ForumType objectType;
	private final List<Article> articles;
	private final LayoutInflater inflater;

	public ThreadRecyclerViewAdapter(Context context, ThreadSafeResponse thread, int forumId, String forumTitle, int objectId, String objectName, ForumType objectType) {
		threadId = thread.getThreadId();
		threadSubject = thread.getThreadSubject();
		this.forumId = forumId;
		this.forumTitle = forumTitle;
		this.objectId = objectId;
		this.objectName = objectName;
		this.objectType = objectType;
		articles = thread.getArticles();
		inflater = LayoutInflater.from(context);
		setHasStableIds(true);
	}

	@NotNull
	@Override
	public ArticleViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
		RowThreadArticleBinding binding = RowThreadArticleBinding.inflate(inflater, parent, false);
		return new ArticleViewHolder(binding);
	}

	@Override
	public void onBindViewHolder(@NotNull ArticleViewHolder holder, int position) {
		holder.bind(articles.get(position));
	}

	@Override
	public int getItemCount() {
		return articles == null ? 0 : articles.size();
	}

	@Override
	public long getItemId(int position) {
		Article article = articles.get(position);
		if (article == null) return RecyclerView.NO_ID;
		return article.getId();
	}

	public int getPosition(int articleId) {
		if (articles == null) return RecyclerView.NO_POSITION;
		for (int i = 0; i < articles.size(); i++) {
			if (articles.get(i).getId() == articleId) return i;
		}
		return RecyclerView.NO_POSITION;
	}

	public class ArticleViewHolder extends RecyclerView.ViewHolder {
		private final RowThreadArticleBinding binding;

		public ArticleViewHolder(RowThreadArticleBinding binding) {
			super(binding.getRoot());
			this.binding = binding;
		}

		public void bind(final Article article) {
			if (article == null) return;

			if (article.getPostTicks() > 0L) {
				binding.rowHeader.setVisibility(View.VISIBLE);
				if (TextUtils.isEmpty(article.getUsername())) {
					binding.username.setVisibility(View.GONE);
				} else {
					binding.username.setText(article.getUsername());
					binding.username.setVisibility(View.VISIBLE);
				}
				binding.postDate.setTimestamp(article.getPostTicks());
				if (article.getEditTicks() != article.getPostTicks()) {
					binding.editDate.setTimestamp(article.getEditTicks());
					binding.editDate.setVisibility(View.VISIBLE);
					binding.dateDivider.setVisibility(View.VISIBLE);
				} else {
					binding.editDate.setVisibility(View.GONE);
					binding.dateDivider.setVisibility(View.GONE);
				}
			} else {
				binding.rowHeader.setVisibility(View.GONE);
			}
			TextViewUtils.setTextMaybeHtml(binding.body, article.getBody().trim());
			binding.viewButton.setOnClickListener(v -> ArticleActivity.start(v.getContext(), threadId, threadSubject, forumId, forumTitle, objectId, objectName, objectType, article));
		}
	}
}

File: app/src/main/java/com/boardgamegeek/ui/adapter/ThreadRecyclerViewAdapter.kt
package com.boardgamegeek.ui.adapter

import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.boardgamegeek.R
import com.boardgamegeek.databinding.RowThreadArticleBinding
import com.boardgamegeek.model.Article
import com.boardgamegeek.model.Forum
import com.boardgamegeek.extensions.inflate
import com.boardgamegeek.extensions.setTextMaybeHtml
import com.boardgamegeek.provider.BggContract
import com.boardgamegeek.ui.ArticleActivity
import com.boardgamegeek.ui.adapter.ThreadRecyclerViewAdapter.ArticleViewHolder
import com.boardgamegeek.util.XmlApi2TagHandler
import kotlin.properties.Delegates

class ThreadRecyclerViewAdapter(
    private val forumId: Int,
    private val forumTitle: String,
    private val objectId: Int,
    private val objectName: String,
    private val objectType: Forum.Type,
) : RecyclerView.Adapter<ArticleViewHolder>(), AutoUpdatableAdapter {

    var threadId: Int = BggContract.INVALID_ID
    var threadSubject: String = ""
    val tagHandler = XmlApi2TagHandler()

    var articles: List<Article> by Delegates.observable(emptyList()) { _, old, new ->
        autoNotify(old, new) { o, n ->
            o == n
        }
    }

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArticleViewHolder {
        return ArticleViewHolder(parent.inflate(R.layout.row_thread_article))
    }

    override fun onBindViewHolder(holder: ArticleViewHolder, position: Int) {
        holder.bind(articles.getOrNull(position))
    }

    override fun getItemCount() = articles.size

    override fun getItemId(position: Int): Long {
        return articles.getOrNull(position)?.id?.toLong() ?: RecyclerView.NO_ID
    }

    fun getPosition(articleId: Int): Int {
        return articles.indexOfFirst { it.id == articleId }
    }

    inner class ArticleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val binding = RowThreadArticleBinding.bind(itemView)

        fun bind(article: Article?) {
            article?.let {
                if (it.postTicks > 0L) {
                    binding.rowHeaderView.isVisible = true
                    binding.usernameView.text = it.username
                    binding.usernameView.isVisible = it.username.isNotBlank()
                    binding.postDateView.timestamp = it.postTicks
                    binding.editDateView.timestamp = it.editTicks
                    binding.editDateView.isVisible = it.editTicks != it.postTicks
                    binding.dateDivider.isVisible = it.editTicks != it.postTicks
                } else {
                    binding.rowHeaderView.isVisible = false
                }
                binding.bodyView.setTextMaybeHtml(it.body.trim(), tagHandler = tagHandler)
                binding.viewButton.setOnClickListener { v: View ->
                    ArticleActivity.start(v.context, threadId, threadSubject, forumId, forumTitle, objectId, objectName, objectType, it)
                }
            }
        }
    }
}
